src/main/drivers/pwm_output_digital.c

```diff
@@ -0,0 +1,239 @@
+/*
+ * This file is part of Betaflight.
+ *
+ * Betaflight is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Betaflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Betaflight. If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <stdbool.h>
+#include <stdint.h>
+#include <math.h>
+
+#include "platform.h"
+
+#include "io.h"
+#include "timer.h"
+#include "pwm_mapping.h"
+#include "pwm_output.h"
+#include "nvic.h"
+#include "dma.h"
+#include "system.h"
+#include "rcc.h"
+
+#define MOTOR_DMA_BUFFER_SIZE 12 /* resolution + frame reset (2us) */
+#define MOTOR_BIT_0 50
+#define MOTOR_BIT_1 117
+
+typedef struct {
+    ioTag_t ioTag;
+    const timerHardware_t *timerHardware;
+    uint16_t value;
+    uint16_t timerDmaSource;
+    uint8_t transferInProgress;
+#if defined(STM32F4)
+    uint32_t dmaBuffer[MOTOR_DMA_BUFFER_SIZE];
+#else
+    uint8_t dmaBuffer[MOTOR_DMA_BUFFER_SIZE];
+#endif
+} motorDmaOutput_t;
+
+static motorDmaOutput_t dmaMotors[MAX_PWM_MOTORS];
+
+bool isTimerConfigured(TIM_TypeDef *timer, uint8_t motorIndex)
+{
+    for (int i = motorIndex - 1; i >= 0; i--) {
+        if (dmaMotors[i].timerHardware && dmaMotors[i].timerHardware->tim == timer) {
+            return true;
+        }
+    }
+    return false;
+}
+
+void pwmWriteDigital(uint8_t index, uint16_t value)
+{
+    uint16_t outputValue = value - 1000;
+    motorDmaOutput_t motor = dmaMotors[index];
+    if (motor.value != outputValue) {
+        for (uint8_t bufferIndex = 0; bufferIndex < MOTOR_DMA_BUFFER_SIZE; bufferIndex++) {
+            motor.dmaBuffer[bufferIndex] = ((outputValue << bufferIndex) & 1) ? MOTOR_BIT_1 : MOTOR_BIT_0;
+        }
+        motor.value = outputValue;        
+    }
+    
+    DMA_SetCurrDataCounter(motor.timerHardware->dmaStream, MOTOR_DMA_BUFFER_SIZE);  
+    DMA_Cmd(motor.timerHardware->dmaStream, ENABLE);
+
+    /* need to check used timers, and only reset once all updates for a given timer are completed */
+    TIM_SetCounter(motor.timerHardware->tim, 0);
+    TIM_DMACmd(motor.timerHardware->tim, motor.timerDmaSource, ENABLE);
+}
+
+static void motor_DMA_IRQHandler(dmaChannelDescriptor_t *descriptor)
+{
+    if (DMA_GET_FLAG_STATUS(descriptor, DMA_IT_TCIF)) {
+        motorDmaOutput_t motor = dmaMotors[descriptor->userParam];
+        motor.transferInProgress = 0;
```

Comment:
fixed this one up - needs to move to some form of double buffering or the like to handle motor updates whilst a transfer is in progress